pkg/blockdevice/v1alpha2/blockdevice.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2019 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	ndmapisv1alpha1 "github.com/openebs/maya/pkg/apis/openebs.io/ndm/v1alpha1"
+	apisv1alpha1 "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+)
+
+//TODO: While using these packages UnitTest must be written to corresponding function
+
+// BlockDeviceState is label for block device states
+type BlockDeviceState string
+
+const (
+	// BlockDeviceStateActive is active state of the block device
+	BlockDeviceStateActive BlockDeviceState = "Active"
+)
+
+// DefaultBlockDeviceCount is a map containing the default block device count of various raid types.
+var DefaultBlockDeviceCount = map[string]int{
+	string(apisv1alpha1.PoolTypeMirroredCPV): int(apisv1alpha1.MirroredBlockDeviceCountCPV),
+	string(apisv1alpha1.PoolTypeStripedCPV):  int(apisv1alpha1.StripedBlockDeviceCountCPV),
+	string(apisv1alpha1.PoolTypeRaidzCPV):    int(apisv1alpha1.RaidzBlockDeviceCountCPV),
+	string(apisv1alpha1.PoolTypeRaidz2CPV):   int(apisv1alpha1.Raidz2BlockDeviceCountCPV),
+}
+
+// BlockDevice encapsulates BlockDevice api object.
+type BlockDevice struct {
+	// actual block device object
+	Object *ndmapisv1alpha1.BlockDevice
+}
+
+// BlockDeviceList holds the list of BlockDevice api
+type BlockDeviceList struct {
+	// list of blockdevices
+	ObjectList *ndmapisv1alpha1.BlockDeviceList
+}
+
+// Predicate defines an abstraction to determine conditional checks against the
+// provided block device instance
+type Predicate func(*BlockDevice) bool
+
+// predicateList holds the list of Predicates
+type predicateList []Predicate
+
+// all returns true if all the predicates succeed against the provided block
+// device instance.
+func (l predicateList) all(c *BlockDevice) bool {
+	for _, pred := range l {
+		if !pred(c) {
+			return false
+		}
+	}
+	return true
+}
+
+// HasAnnotation returns true if provided annotation key and value are present
+// in the provided block deive instance.
+func HasAnnotation(key, value string) Predicate {
```

Comment:
Please expose a method as well.
Please expose corresponding method for each Predicate function.
Predicate func will internally invoke the method

Reply:
Done in all places